### PR TITLE
Add List Versions operation in Remote Config

### PR DESCRIPTION
**Pull request by lahirumaramba**
- Add List Versions operation in Remote Config
- Add  `ListVersionsOptions`, `ListVersionsResult` types
- Add unit tests

Related to #446

### CODE REVIEW

**Comment by lahirumaramba** [comment]
Note: Still in draft. Needs to be rebased after #497

File: src/test/java/com/google/firebase/remoteconfig/FirebaseRemoteConfigClientImplTest.java [hunk 1]
```diff
@@ -914,6 +919,196 @@ public void testRollbackErrorWithRcError() throws IOException {
     }
   }
 
+
```
Comment (hiranya911) [inline src/test/java/com/google/firebase/remoteconfig/FirebaseRemoteConfigClientImplTest.java:922]: Redundant whitespace

File: src/main/java/com/google/firebase/remoteconfig/ListVersionsOptions.java [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+    }
+
+    /**
+     * Sets the earliest update time to include in the results.
+     *
+     * @param startTimeMillis Specify the earliest update time to include in the results.
+     *                        Any entries updated before this time are omitted.
+     * @return This builder.
+     */
+    public Builder setStartTime(long startTimeMillis) {
```
Comment (hiranya911) [inline src/main/java/com/google/firebase/remoteconfig/ListVersionsOptions.java:155]: setStartTimeMillis?

File: src/test/java/com/google/firebase/remoteconfig/ListVersionsResultTest.java [hunk 3]
```diff
@@ -0,0 +1,340 @@
...
+  @Test
+  public void testIteratorWithNoVersions() throws FirebaseRemoteConfigException {
+    ListVersionsResult.VersionsResultBatch result = new ListVersionsResult.VersionsResultBatch(
+            ImmutableList.<Version>of(),
+            ListVersionsResult.END_OF_LIST);
+    TestVersionSource source = new TestVersionSource(result);
+
+    ListVersionsResult page = new ListVersionsResult.Factory(source).create();
+    Iterator<Version> iterator = page.iterateAll().iterator();
+    while (iterator.hasNext()) {
```
Comment (hiranya911) [inline src/test/java/com/google/firebase/remoteconfig/ListVersionsResultTest.java:271]: `assertFalse(iterator.hasNext())`

**Review by hiranya911 — CHANGES_REQUESTED** [review]
Looks pretty good. Just a few thoughts on improving the implementation.

File: src/main/java/com/google/firebase/remoteconfig/ListVersionsOptions.java [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+
+    /**
+     * Sets the newest version number to include in the results.
+     *
+     * @param endVersionNumber Specify the newest version number to include in the results.
+     *                         If specified, must be greater than zero. Defaults to the newest
+     *                         version.
+     * @return This builder.
+     */
+    public Builder setEndVersionNumber(String endVersionNumber) {
```
[the remark this answers is not in the document]
Comment (lahirumaramba) [reply src/main/java/com/google/firebase/remoteconfig/ListVersionsOptions.java:147]: Good catch! Added a setter that takes a `long`

File: src/main/java/com/google/firebase/remoteconfig/ListVersionsOptions.java [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+    }
+
+    /**
+     * Sets the page size.
+     *
+     * @param pageSize The maximum number of items to return per page.
+     * @return This builder.
+     */
+    public Builder setPageSize(int pageSize) {
+      checkArgument(pageSize > 0 && pageSize < 301,
```
[the remark this answers is not in the document]
Comment (lahirumaramba) [reply src/main/java/com/google/firebase/remoteconfig/ListVersionsOptions.java]: Moved the validation to the constructor. Thanks!

File: src/main/java/com/google/firebase/remoteconfig/ListVersionsOptions.java [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+    }
+
+    private String convertToUtcZuluFormat(long millis) {
+      checkArgument(millis >= 0, "Milliseconds duration must not be negative");
+      SimpleDateFormat dateFormat = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS000000'Z'");
+      dateFormat.setTimeZone(TimeZone.getTimeZone("UTC"));
+      return dateFormat.format(new Date(millis));
+    }
+
+    private boolean isValidVersionNumber(String versionNumber) {
```
[the remark this answers is not in the document]
Comment (lahirumaramba) [reply src/main/java/com/google/firebase/remoteconfig/ListVersionsOptions.java]: Good call! Moved the validations to `RemoteConfigUtil`

File: src/main/java/com/google/firebase/remoteconfig/ListVersionsOptions.java [hunk 7]
```diff
@@ -0,0 +1,205 @@
...
+     * @param endTimeMillis Specify the latest update time to include in the results.
+     *                      Any entries updated on or after this time are omitted.
+     * @return This builder.
+     */
+    public Builder setEndTimeMillis(long endTimeMillis) {
+      this.endTime = convertToUtcZuluFormat(endTimeMillis);
+      return this;
+    }
+
+    private String convertToUtcZuluFormat(long millis) {
```
Comment (hiranya911) [inline src/main/java/com/google/firebase/remoteconfig/ListVersionsOptions.java:189]: Probably belongs in the shared util. I assume there are other places where this is needed.

**Review by hiranya911 — APPROVED** [review]
Thanks! LGTM with a suggestion.

File: src/main/java/com/google/firebase/remoteconfig/ListVersionsOptions.java (hunk 7 above)
Comment (lahirumaramba) [reply src/main/java/com/google/firebase/remoteconfig/ListVersionsOptions.java]: Good call. Currently, we use it only in `ListVersionsOptions`. I moved it to Utils anyway to future proof :)

**Pull request merged**